pkg/controller/v1alpha2/core/scopes/healthscope/healthscope.go

```diff
@@ -19,133 +19,198 @@ package healthscope
+		return nil
+	}
+	r := &HealthCondition{
+		IsHealthy: false,
+		Target:    ref,
+	}
+	cwObj := corev1alpha2.ContainerizedWorkload{}
+	cwObj.SetGroupVersionKind(corev1alpha2.SchemeGroupVersion.WithKind(kindContainerizedWorkload))
+	if err := c.Get(ctx, types.NamespacedName{Namespace: namespace, Name: ref.Name}, &cwObj); err != nil {
+		r.Diagnosis = errors.Wrap(err, errHealthCheck).Error()
+		return r
+	}
+	r.Target.UID = cwObj.GetUID()
+
+	r.SubConditions = []*HealthCondition{}
+	childRefs := cwObj.Status.Resources
+
+	for _, childRef := range childRefs {
+		switch childRef.Kind {
+		case kindDeployment:
+			// reuse Deployment health checker
+			childCondition := CheckDeploymentHealth(ctx, c, childRef, namespace)
+			r.SubConditions = append(r.SubConditions, childCondition)
+		default:
+			childCondition := &HealthCondition{
+				Target:    childRef,
+				IsHealthy: true,
+			}
+			o := unstructured.Unstructured{}
+			o.SetAPIVersion(childRef.APIVersion)
+			o.SetKind(childRef.Kind)
+			if err := c.Get(ctx, types.NamespacedName{Namespace: namespace, Name: childRef.Name}, &o); err != nil {
+				// for unspecified resource
+				// if cannot get it, then check fails
+				childCondition.IsHealthy = false
+				childCondition.Diagnosis = errors.Wrap(err, errHealthCheck).Error()
+			}
+			r.SubConditions = append(r.SubConditions, childCondition)
+		}
 	}
 
-	// TODO(artursouza): add other health checks.
-	// Generic health check by validating if the resource exists.
-	object := unstructured.Unstructured{}
-	object.SetAPIVersion(ref.APIVersion)
-	object.SetKind(ref.Kind)
-	objectRef := types.NamespacedName{Namespace: namespace, Name: ref.Name}
-	err := client.Get(ctx, objectRef, &object)
-	return err
+	r.IsHealthy = true
+	for _, sc := range r.SubConditions {
+		if !sc.IsHealthy {
+			r.IsHealthy = false
+			r.Diagnosis = errUnhealthyChildResource
+			break
+		}
+	}
+	return r
 }
 
-func deploymentHealthStatus(ctx context.Context, client client.Client, namespace string, ref runtimev1alpha1.TypedReference) error {
+// CheckDeploymentHealth checks health status of Deployment
+func CheckDeploymentHealth(ctx context.Context, client client.Client, ref runtimev1alpha1.TypedReference, namespace string) *HealthCondition {
+	if ref.GroupVersionKind() != apps.SchemeGroupVersion.WithKind(kindDeployment) {
+		return nil
+	}
+	r := &HealthCondition{
+		IsHealthy: false,
+		Target:    ref,
+	}
 	deployment := apps.Deployment{}
-	deployment.APIVersion = ref.APIVersion
-	deployment.Kind = ref.Kind
+	deployment.SetGroupVersionKind(apps.SchemeGroupVersion.WithKind(kindDeployment))
 	deploymentRef := types.NamespacedName{Namespace: namespace, Name: ref.Name}
 	if err := client.Get(ctx, deploymentRef, &deployment); err != nil {
-		return errors.Wrapf(err, errResourceNotFound, ref.APIVersion, ref.Kind, ref.Name)
+		r.Diagnosis = errors.Wrap(err, errHealthCheck).Error()
+		return r
 	}
+	r.Target.UID = deployment.GetUID()
 
 	if deployment.Status.ReadyReplicas == 0 {
```

Comment:
I may be missing something here, why is that "ReadyReplicas != 0" means healthy?

Reply:
Actually I just stay in line with [current logic on health checking](https://github.com/crossplane/oam-kubernetes-runtime/blob/bc0db9fd48f9d577c6bb4cf6f715cffded6416ed/pkg/controller/v1alpha2/core/scopes/healthscope/healthscope.go#L147) .And a more rigorous check is `x.Status.ReadyReplicas == x.Spec.Replicas`

Reply:
I think we could fix it in the following PRs